packages/tracing/src/idletransaction.ts

```diff
@@ -0,0 +1,284 @@
+// tslint:disable: max-classes-per-file
+import { Hub } from '@sentry/hub';
+import { TransactionContext } from '@sentry/types';
+import { logger, timestampWithMs } from '@sentry/utils';
+
+import { Span } from './span';
+import { SpanStatus } from './spanstatus';
+import { SpanRecorder, Transaction } from './transaction';
+
+const DEFAULT_IDLE_TIMEOUT = 1000;
+
+/**
+ * @inheritDoc
+ */
+export class IdleTransactionSpanRecorder extends SpanRecorder {
+  private readonly _pushActivity: (id: string) => void;
+  private readonly _popActivity: (id: string) => void;
+  public transactionSpanId: string = '';
+
+  public constructor(
+    pushActivity: (id: string) => void,
+    popActivity: (id: string) => void,
+    transactionSpanId: string = '',
+    maxlen?: number,
+  ) {
+    super(maxlen);
+    this._pushActivity = pushActivity;
+    this._popActivity = popActivity;
+    this.transactionSpanId = transactionSpanId;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public add(span: Span): void {
+    // We should make sure we do not push and pop activities for
+    // the transaction that this span recorder belongs to.
+    if (span.spanId !== this.transactionSpanId) {
+      // We patch span.finish() to pop an activity after setting an endTimestamp.
+      span.finish = (endTimestamp?: number) => {
+        span.endTimestamp = typeof endTimestamp === 'number' ? endTimestamp : timestampWithMs();
+        this._popActivity(span.spanId);
+      };
+
+      // We should only push new activities if the span does not have an end timestamp.
+      if (span.endTimestamp === undefined) {
+        this._pushActivity(span.spanId);
+      }
+    }
+
+    super.add(span);
+  }
+}
+
+/**
+ * An IdleTransaction is a transaction that automatically finishes. It does this by tracking child spans as activities.
+ * You can have multiple IdleTransactions active, but if the `onScope` option is specified, the idle transaction will
+ * put itself on the scope on creation.
+ */
+export class IdleTransaction extends Transaction {
+  // Activities store a list of active spans
+  // TODO: Can we use `Set()` here?
+  public activities: Record<string, boolean> = {};
```

Comment:
@kamilogorek can we use es6 `Set()` in this codebase? - or do you think we should stay away from this and just use regular JS objects.

Reply:
Let's use object, `Set` is only support IE 11